go/tools/builders/compilepkg.go

```diff
@@ -348,23 +318,26 @@ func compileArchive(
 	// C files.
 	var objFiles []string
 	if compilingWithCgo {
-		// If the package uses Cgo, compile .s and .S files with cgo2, not the Go assembler.
-		// Otherwise: the .s/.S files will be compiled with the Go assembler later
 		var srcDir string
-		srcDir, goSrcs, objFiles, err = cgo2(goenv, goSrcs, cgoSrcs, cSrcs, cxxSrcs, objcSrcs, objcxxSrcs, sSrcs, hSrcs, packagePath, packageName, cc, cppFlags, cFlags, cxxFlags, objcFlags, objcxxFlags, ldFlags, cgoExportHPath)
-		if err != nil {
-			return err
-		}
-		if coverMode != "" && nogoPath != "" {
-			// Compile original source files, not coverage instrumented, for nogo
-			_, goSrcsNogo, _, err = cgo2(goenv, goSrcsNogo, cgoSrcsNogo, cSrcs, cxxSrcs, objcSrcs, objcxxSrcs, sSrcs, hSrcs, packagePath, packageName, cc, cppFlags, cFlags, cxxFlags, objcFlags, objcxxFlags, ldFlags, cgoExportHPath)
+		// Also run cgo on original source files, not coverage instrumented, if
+		// using nogo.
+		if coverMode != "" && cgoGoSrcsForNogoPath != "" {
+			srcDir, goSrcs, objFiles, err = cgo2(goenv, goSrcs, cgoSrcs, cSrcs, cxxSrcs, objcSrcs, objcxxSrcs, sSrcs, hSrcs, packagePath, packageName, cc, cppFlags, cFlags, cxxFlags, objcFlags, objcxxFlags, ldFlags, cgoExportHPath, "")
+			if err != nil {
+				return err
+			}
+			_, _, _, err = cgo2(goenv, goSrcsNogo, cgoSrcsNogo, cSrcs, cxxSrcs, objcSrcs, objcxxSrcs, sSrcs, hSrcs, packagePath, packageName, cc, cppFlags, cFlags, cxxFlags, objcFlags, objcxxFlags, ldFlags, cgoExportHPath, cgoGoSrcsForNogoPath)
 			if err != nil {
 				return err
 			}
 		} else {
-			goSrcsNogo = goSrcs
+			// If the package uses Cgo, compile .s and .S files with cgo2, not the Go assembler.
+			// Otherwise: the .s/.S files will be compiled with the Go assembler later
+			srcDir, goSrcs, objFiles, err = cgo2(goenv, goSrcs, cgoSrcs, cSrcs, cxxSrcs, objcSrcs, objcxxSrcs, sSrcs, hSrcs, packagePath, packageName, cc, cppFlags, cFlags, cxxFlags, objcFlags, objcxxFlags, ldFlags, cgoExportHPath, cgoGoSrcsForNogoPath)
```

Comment:
Nit: the comment is supposed to apply to both the if and else clauses, not exclusively to the else clause.

Perhaps we could move this to outside of the else clause like this

```
cgoSrcsForNogo := cgoGoSrcsForNogoPath
if coverMode != "" && cgoGoSrcsForNogoPath != "" {
  cgoSrcsForNogo = ""
  ...
}
srcDir, goSrcs, objFiles, err = cgo2(..., cgoSrcsForNogo)
...
```

wdyt?

Reply:
I thought about doing that, but failed to find a layout that communicates that we are running `cgo` twice and ignoring its outputs in such a clear way. I added a TODO about getting rid of the unnecessarily duplicated compilation and also duplicated the comment on when to use which assembler.

Reply:
Nvm about the TODO, cgo needs the compilation output to generate code.

Reply:
Here is an example cgo regex we have used to exclude out nogo failures on generated code 

https://github.com/buildbuddy-io/buildbuddy/blob/d8461cff320849f1f9dc8aef322a15a525f76f10/BUILD#L39-L41

I don't mind using a different regex, but ideally, there should be a regex that we could reliably exclude generated code.